Incremental Configuration Updates

oxzi opened the pull request:
Previously, the entire configuration stored in the database was synchronized every second. With the growth of configurations in live environments on the horizon, this would simply not scale well. This brings us to incremental updates.

By introducing two new columns - "changed_at" as a Unix millisecond timestamp and "deleted" as a boolean - for all tables referenced in the ConfigSet structure, SQL queries can be modified to retrieve only those rows with a more recent timestamp. The "deleted" column became necessary to detect disappearances, since the synchronization now only takes newer items into account. Some additional fields needed to be added to the ConfigSet to track relationships.

Even though the codebase served well at the time, there was some code that did almost the same thing as other code, just in different ways.  So a huge refactoring was done. This resulted in an internal generic function that handles all synchronization with custom callbacks.

The web counterpart is being developed in <https://github.com/Icinga/icinga-notifications-web/pull/187>.

Closes #5.

oxzi commented:
After talking with @sukhwinder33445 about this change as he implements the web counterpart, I have removed the `incr_cfg_delete` function and its trigger usage, rewriting a `DELETE` to an `UPDATE` setting `deleted = 'y'` on the main configuration tables.

For historical reasons and if it might become relevant again, I have attached the diff below.

<details>
<summary>incr_cfg_delete diff</summary>

```diff
diff --git a/schema/pgsql/schema.sql b/schema/pgsql/schema.sql
index a5abfea..08f5406 100644
--- a/schema/pgsql/schema.sql
+++ b/schema/pgsql/schema.sql
@@ -88,6 +88,23 @@ CREATE FUNCTION incr_cfg_bump_changed_at_relation_2nd() -- join_tbl text, foreig
     END;
     $$;
 
+-- incr_cfg_delete is a BEFORE TRIGGER for DELETE, which sets changed_at to the current timestamp and raises the deleted
+-- column instead of actually deleting the row.
+CREATE FUNCTION incr_cfg_delete()
+    RETURNS trigger
+    LANGUAGE plpgsql
+    AS $$
+    BEGIN
+        -- Cannot change OLD's attributes, as this would require returning OLD which implies continuing with deletion.
+        EXECUTE format('
+            UPDATE %s
+            SET changed_at = EXTRACT(EPOCH FROM NOW()) * 1000, deleted = ''y''
+            WHERE id = %s',
+            TG_TABLE_NAME, OLD.id);
+        RETURN NULL;
+    END;
+    $$;
+
 CREATE TABLE available_channel_type (
     type text NOT NULL,
     name text NOT NULL,
@@ -119,6 +136,11 @@ CREATE TRIGGER trg_channel_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_channel_incr_cfg_delete
+    BEFORE DELETE ON channel
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TABLE contact (
     id bigserial,
     full_name text NOT NULL,
@@ -140,6 +162,11 @@ CREATE TRIGGER trg_contact_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contact_incr_cfg_delete
+    BEFORE DELETE ON contact
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TABLE contact_address (
     id bigserial,
     contact_id bigint NOT NULL REFERENCES contact(id),
@@ -160,6 +187,11 @@ CREATE TRIGGER trg_contact_address_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contact_address_incr_cfg_delete
+    BEFORE DELETE ON contact_address
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TABLE contactgroup (
     id bigserial,
     name text NOT NULL,
@@ -178,6 +210,11 @@ CREATE TRIGGER trg_contactgroup_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contactgroup_incr_cfg_delete
+    BEFORE DELETE ON contactgroup
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 -- Changes to contactgroup_member should be notified by an updated contactgroup.changed_at.
 CREATE TABLE contactgroup_member (
     contactgroup_id bigint NOT NULL REFERENCES contactgroup(id),
@@ -208,6 +245,11 @@ CREATE TRIGGER trg_schedule_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_schedule_incr_cfg_delete
+    BEFORE DELETE ON schedule
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TABLE timeperiod (
     id bigserial,
     owned_by_schedule_id bigint REFERENCES schedule(id), -- nullable for future standalone timeperiods
@@ -225,6 +267,11 @@ CREATE TRIGGER trg_timeperiod_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_timeperiod_incr_cfg_delete
+    BEFORE DELETE ON timeperiod
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 -- Changes to timeperiod_entry should be notified by an updated timeperiod.changed_at.
 CREATE TABLE timeperiod_entry (
     id bigserial,
@@ -315,6 +362,11 @@ CREATE TRIGGER trg_source_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_source_incr_cfg_delete
+    BEFORE DELETE ON source
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TABLE object (
     id bytea NOT NULL, -- SHA256 of identifying tags and the source.id
     source_id bigint NOT NULL REFERENCES source(id),
@@ -377,6 +429,11 @@ CREATE TRIGGER trg_rule_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_rule_incr_cfg_delete
+    BEFORE DELETE ON rule
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 -- Changes to rule_escalation should be notified by an updated rule.changed_at.
 CREATE TABLE rule_escalation (
     id bigserial,
diff --git a/schema/pgsql/upgrades/025.sql b/schema/pgsql/upgrades/025.sql
index ee5f595..3c36de4 100644
--- a/schema/pgsql/upgrades/025.sql
+++ b/schema/pgsql/upgrades/025.sql
@@ -52,6 +52,21 @@ CREATE FUNCTION incr_cfg_bump_changed_at_relation_2nd() -- join_tbl text, foreig
     END;
     $$;
 
+CREATE FUNCTION incr_cfg_delete()
+    RETURNS trigger
+    LANGUAGE plpgsql
+    AS $$
+    BEGIN
+        -- Cannot change OLD's attributes, as this would require returning OLD which implies continuing with deletion.
+        EXECUTE format('
+                UPDATE %s
+                SET changed_at = EXTRACT(EPOCH FROM NOW()) * 1000, deleted = ''y''
+                WHERE id = %s',
+                       TG_TABLE_NAME, OLD.id);
+        RETURN NULL;
+    END;
+    $$;
+
 ALTER TABLE channel
     ADD COLUMN changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
     ADD COLUMN deleted boolenum NOT NULL DEFAULT 'n';
@@ -98,21 +113,41 @@ CREATE TRIGGER trg_channel_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_channel_incr_cfg_delete
+    BEFORE DELETE ON channel
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_contact_incr_cfg_update
     BEFORE INSERT OR UPDATE ON contact
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contact_incr_cfg_delete
+    BEFORE DELETE ON contact
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_contact_address_incr_cfg_update
     BEFORE INSERT OR UPDATE ON contact_address
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contact_address_incr_cfg_delete
+    BEFORE DELETE ON contact_address
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_contactgroup_incr_cfg_update
     BEFORE INSERT OR UPDATE ON contactgroup
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_contactgroup_incr_cfg_delete
+    BEFORE DELETE ON contactgroup
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_contactgroup_member_changed_at_relation
     AFTER INSERT OR UPDATE OR DELETE ON contactgroup_member
     FOR EACH ROW
@@ -123,11 +158,21 @@ CREATE TRIGGER trg_schedule_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_schedule_incr_cfg_delete
+    BEFORE DELETE ON schedule
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_timeperiod_incr_cfg_update
     BEFORE INSERT OR UPDATE ON timeperiod
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_timeperiod_incr_cfg_delete
+    BEFORE DELETE ON timeperiod
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_timeperiod_entry_changed_at_relation
     AFTER INSERT OR UPDATE OR DELETE ON timeperiod_entry
     FOR EACH ROW
@@ -143,11 +188,21 @@ CREATE TRIGGER trg_source_incr_cfg_update
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_source_incr_cfg_delete
+    BEFORE DELETE ON source
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_rule_incr_cfg_update
     BEFORE INSERT OR UPDATE ON rule
     FOR EACH ROW
     EXECUTE FUNCTION incr_cfg_bump_changed_at();
 
+CREATE TRIGGER trg_rule_incr_cfg_delete
+    BEFORE DELETE ON rule
+    FOR EACH ROW
+    EXECUTE FUNCTION incr_cfg_delete();
+
 CREATE TRIGGER trg_rule_escalation_changed_at_relation
     AFTER INSERT OR UPDATE OR DELETE ON rule_escalation
     FOR EACH ROW
```
</details>

julianbrost commented:
My thinking when suggesting to not have `changed_at`/`deleted` in every config table but only in the main ones was a bit that should also be the natural thing to do in Web. For example, if you add an recipient somewhere in an escalation, you do this on the edit event rule page and you click save there, so there it would only be necessary to update `rule.changed_at` when the form is saved, all the changes to other tables would be covered implicitly by that.

Now this turned in to quite a bit of complex trigger mechanics instead. I would definitely consider adding these two columns to all tables instead of requiring three variants of the trigger functions. Is everything you implemented in PostgreSQL triggers also possible with MySQL and MariaDB?

> as well as raising the `deleted` column instead of actually performing the `DELETE` command.

Does raising have any special meaning in this context? Or is it simply supposed to say `SET deleted = 'y'`?

Also, I'm not yet sure how good the idea is to hook `DELETE` in a way that it doesn't actually delete, i.e. might this be too surprising so that this could result in unexpected behavior? The only downside I see with executing an `UPDATE ... SET deleted = 'y'` instead would be that with the current current distinction into main/child config tables, you'd have to make a difference between those. If we'd decide to add the columns to every table, this would no longer be the case. (But on the other hand, there would then be a difference between adding a new group membership and reviving a previously deleted one.)

sukhwinder33445 commented:
I would say that if for example the `rule` is updated via `form` (rule.name ="UpdatedName"), the `changed_at` column should be updated automatically. Maridb/Mysql can do this with [ON UPDATE](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html). If an INSERT/UPDATE/DELETE takes place in the related (child) tables, the web should (can) update the `changed_at` column of the main(parent) table manually.

julianbrost commented:
> I would say that if for example the `rule` is updated via `form` (rule.name ="UpdatedName"), the `changed_at` column should be updated automatically.

> If an INSERT/UPDATE/DELETE takes place in the related (child) tables, the web should (can) update the `changed_at` column of the main(parent) table manually.

Why do you want to make a distinction here? Would that make anything easier for web?

> Maridb/Mysql can do this with [ON UPDATE](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html).

Does this even work here? Reads like this is a magic feature just for `TIMESTAMP` and `DATETIME` which we don't use.

sukhwinder33445 commented:
> Why do you want to make a distinction here? Would that make anything easier for web?

Yes. For example, if the rule name is changed, the form returns new values that we can write directly to `$db->update('rule', $formValues);`. But now $formValues also needs the column `changed_at`. Adding this column is no big deal, but if db can do it itself, why should we do it manually. This will also prevent future errors if someone forgets to add this column to the $formValues.

> Does this even work here? Reads like this is a magic feature just for TIMESTAMP and DATETIME which we don't use.

But `changed_at` is handled as `timestamp`.

Update:
I thought it would be simpler than the changes on the web, but it's not, and it's getting more complex with the triggers. And the data type is not a real `timestamp`. I can understand your reasoning now.

yhabteab reviewed (commented):
Just a few comments that caught my eyes!

oxzi commented:
Small note, mostly for future me: I have just rebased against the current `main`, nothing more. I am going to address @yhabteab's comments later. Furthermore, I talked with @julianbrost about the `ConfigSet`, which could contain every table/relation instead of groups of relations, easing fetching logic and allowing more unification.

oxzi commented:
TODO: Fetch each table by its own. This "flattening" should ease the synchronization and might even allow more code reusage.

julianbrost commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,207 @@
+	dest interface{},
+) error {
+	stmt := r.db.BuildSelectStmt(typePtr, typePtr) + ` WHERE "changed_at" > COALESCE(?, CAST(0 AS BIGINT))`
+	if len(ids) > 0 {
+		stmt += ` OR "` + idField + `" IN (` + strings.Join(strings.Split(strings.Repeat("?", len(ids)), ""), ",") + `)`
+	}
+	stmt += ` ORDER BY "changed_at"`
+	stmt = r.db.Rebind(stmt)
+
+	args := make([]interface{}, 0, 1+len(ids))
+	args = append(args, changedAt)
+	for id := range ids {
+		args = append(args, id)
+	}
```
`sqlx.In()` should do most of this for you repeating placeholders and flattening/expanding the arguments.

oxzi replied:
This function was dead code and could be removed.
See also https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1613701885

sukhwinder33445 commented:
Are the `sql` files missing by mistake, or have you decided for a new implementation?
I assume that all tables mentioned [here](https://github.com/Icinga/icinga-notifications/issues/5#issuecomment-2077030663) get the `deleted` and `changed_at` flag (as in the recently removed `sql` file). 
The implementation in the web is based on this assumption.

oxzi commented:
@sukhwinder33445: Sorry, the schema and an update are now part of the work in progress commit.

sukhwinder33445 commented on schema/pgsql/upgrades/031.sql:
```diff
@@ -0,0 +1,83 @@
+
+CREATE INDEX idx_contactgroup_changed_at ON contactgroup(changed_at);
+
+ALTER TABLE contacgroup_member
+    ADD COLUMN changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    ADD COLUMN deleted boolenum NOT NULL DEFAULT 'n';
+
+CREATE INDEX idx_contacgroup_member_changed_at ON contacgroup_member(changed_at);
```
You have a typo here, `contacgroup_member` -> `contactgroup_member`.

oxzi replied:
Oops. Thanks.

oxzi commented on internal/config/utils.go:
```diff
@@ -0,0 +1,58 @@
+	args = append(args, changedAt)
+	for id := range ids {
+		args = append(args, id)
+	}
```
This method was lost along the refactoring way. But thanks anyway :)

oxzi commented on internal/config/rule.go:
```diff
@@ -5,58 +5,85 @@ import (
+		tx,
+		escalationPtr,
+		"rule_id",
+		utils.MapKeys(rulesByID),
```
This is now also gone.

oxzi commented on internal/channel/channel.go:
```diff
@@ -27,6 +32,31 @@ type Channel struct {
+	return nil
+}
+
+func (c *Channel) GetID() int64 {
+	return c.ID
+}
+
+func (c *Channel) GetChangedAt() types.UnixMilli {
+	return c.ChangedAt
+}
+
+func (c *Channel) IsDeleted() bool {
+	return c.Deleted.Valid && c.Deleted.Bool
+}
```
I thought about adding some `struct` type to be embedded elsewhere, but this would lead to cyclic imports unless I would locate this new type somewhere new. Tomorrow, I am going to play around with this idea a bit more.

oxzi replied:
I broke this down even further by splitting the `Validate` method off into its own interface, as almost no type required an implementation therefore.

https://github.com/Icinga/icinga-notifications/blob/b746738576303c41e43789e36fed6e5295848f45/internal/config/incremental_sync.go#L28

oxzi replied:
Thanks for this suggestion, @yhabteab. I have given this a try in my implementation.

oxzi commented on internal/config/channel.go:
```diff
@@ -8,79 +8,23 @@ import (
+			element.Start(context.TODO(), r.logs.GetChildLogger("channel").With(
+				zap.Int64("id", element.ID),
+				zap.String("name", element.Name)))
+		})
```
For the moment, I went with option 1 and gave all those callback hook functions describing parameter names.

julianbrost commented on internal/recipient/group.go:
```diff
@@ -32,4 +48,37 @@ func (g *Group) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
 }
 
+type GroupMember struct {
+	GroupId   int64 `db:"contactgroup_id"`
+	ContactId int64 `db:"contact_id"`
```
Just an idea, not yet sure if this will be nicer in the end: Move these two fields into a new type `GroupMemberKey`, embed it here and then use it instead of `CompTuple[int64, int64]`. That would simplify the `GetID()` function (not that it's very complex, but still it would take it from simple to trivial), remove any ambiguity which ID is `A` and which is `B`. And also, it seems to be the only use of `CompTuple`, so that would no longer be necessary.[^1]

[^1]: And removing that type will have the additional benefit that it stops me from making stupid suggestions like renaming it to `ComPair` :zany_face:

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,228 @@
+			logger.Info("Deleted configuration element")
+		} else if oldExists {
+			logger = logger.With(zap.Object("old", oldT), zap.Object("update", newT))
+			if updateFn != nil {
+				if err := updateFn(oldT, newT); errors.Is(err, reAddUpdateFnErr) {
+					logger.Debug("Invoking update by removing and re-adding element")
+					if err := deleteFn(oldT); err != nil {
+						logger.Errorw("Updating requested deleting first, which failed", zap.Error(err))
+						countErr++
+						continue
+					}
+					if err := createFn(newT); err != nil {
+						logger.Errorw("Updating requested re-creating, which failed", zap.Error(err))
+						countErr++
+						continue
+					}
+				} else if err != nil {
+					logger.Errorw("Updating callback failed", zap.Error(err))
+					countErr++
+					continue
+				}
+			}
```
Valid point, but what exactly should be done if nothing is specified? Just replace the old element with the new?

There is currently no `incrementalApplyPending` call with a `nil` value for `updateFn`.

What would you suggest?
1. Re-add the element and optionally call the other two callbacks, if available.
2. Raise an error.
3. Do nothing (current situation; kinda foolish).
4. Something else.

julianbrost replied:
Good question, I haven't looked through what all the callbacks do in detail so far. One thing I noticed so far is that you have four instances where this callback is passed, `updateFn == nil` could be made to mean just that: https://github.com/Icinga/icinga-notifications/blob/90f793f2bdf4fd9bb695b779be4799fcfbcbb38f/internal/config/rule.go#L95-L97

Oh and one thing I forgot to ask also concerning this code block: I'm wondering why (at least) the `reAddUpdateFnErr` case doesn't do anything like `delete(*configSetField, id)` and `(*configSetField)[id] = newT` which is done in the regular add and remove cases. Most of the callbacks seem to copy values from `newT` to `oldT` which doesn't happen in the `reAddUpdateFnErr` case.

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,228 @@
+
+	stmt := r.db.BuildSelectStmt(typePtr, typePtr)
+	stmtArgs := []any{}
+	if hasChangedAt {
+		stmtLogger = stmtLogger.With(zap.Time("changed_at", changedAt.Time()))
+		stmt += ` WHERE "changed_at" > ?`
+		stmtArgs = []any{changedAt}
+	}
```
Yes, both for the initial loading as well as any subsequent loading if nothing was fetched during the first (or other previous) runs.

That's a good idea. In the current design, initially loaded deleted elements would result in a warning.
https://github.com/Icinga/icinga-notifications/blob/90f793f2bdf4fd9bb695b779be4799fcfbcbb38f/internal/config/incremental_sync.go#L161-L163
By doing so, this would be circumvented.

julianbrost reviewed (commented):
For the moment, two comments on the recent changes to the schema:

oxzi commented on schema/pgsql/schema.sql:
```diff
@@ -48,55 +48,87 @@ CREATE TABLE channel (
     id bigserial,
     full_name citext NOT NULL,
-    username citext, -- reference to web user
+    -- Reference web user.
+    -- Must be NULLed for deletion via "deleted = 'y'", full_name should be altered to contain the old username.
+    username citext,
```
The idea behind this one was to allow displaying deleted users in the history tables. As, by a bit naïve assumption, the `username` might refer to the _real_ username, the `full_name` might become unclear later on.

This resulted as an idea from a discussion with @sukhwinder33445 and @nilmerg, but currently the way after this PR is, afaik, still unclear. Should it even be possible to have this "ghost" entries being shown - as, i.e., a deleted contact - or should a garbage collection be created or something else?

oxzi replied:
I removed the comment from the schema as it does not really belong there.

yhabteab commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,247 @@
+			zap.String("table", tableName),
+			zap.Any("id", id))
+
+		createAction := func() error {
+			if createFn != nil {
+				if err := createFn(newT); err != nil {
+					countErr++
+					return fmt.Errorf("creation callback error, %w", err)
+				}
+			}
+			(*configSetField)[id] = newT
+			countNew++
+			return nil
+		}
+
+		deleteAction := func() error {
+			defer delete(*configSetField, id)
+			countDel++
+			if deleteFn != nil {
+				if err := deleteFn(oldT); err != nil {
+					countErr++
+					return fmt.Errorf("deletion callback error, %w", err)
+				}
+			}
+			return nil
+		}
```
You can parameterise these functions and declare them outside the loop.

yhabteab commented on internal/config/rule.go:
```diff
@@ -1,195 +1,112 @@
+		func(curElement, update *rule.Rule) error {
+			curElement.IsActive = update.IsActive
+			curElement.Name = update.Name
+			curElement.ObjectFilter = update.ObjectFilter
```
Can't `rule#TimePeriodID` also be changed at runtime?

oxzi replied:
Good catch!

yhabteab reviewed (commented):
You have defined json tags for all `ChangedAt` and `Deleted` fields, but seem to ignore them completely in all update function hooks. Is that intentional? If so, then the json tags are probably useless.

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,247 @@
+		return err
+	}
+
+	*changeConfigSetField = make(map[PK]T)
```
I tried this, but found the current way more readable.

yhabteab commented on internal/config/rule.go:
```diff
@@ -1,195 +1,126 @@
+			curElement.Name = update.Name
+			curElement.ObjectFilter = update.ObjectFilter
+
+			if curElement.TimePeriodID != update.TimePeriodID {
+				if update.TimePeriodID.Valid {
```
You must also update the `curElement.TimePeriodID` and you're also still missing `ObjectFilterExpr` here.

oxzi replied:
So far I have skipped fields only having a value within the database and being used only once to populate some Go field as, .e.g, most `…Expr` fields or `…Id` fields.

However, this leads to "inconsequential" structs, potentially leading to confusions when being dumped as JSON.
I'll fix this.

yhabteab commented on internal/config/rule.go:
```diff
@@ -1,195 +1,126 @@
+			return nil
+		},
+		func(curElement, update *rule.Escalation) error {
+			curElement.NameRaw = update.NameRaw
+			curElement.Condition = update.Condition
```
You're updating the parsed escalation condition but forgot to also update the actual filter string in `ConditionExpr`?

oxzi replied:
https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1660670544

yhabteab commented on internal/config/rule.go:
```diff
@@ -1,195 +1,126 @@
+			ruleFound := false
+			for id, elementRule := range r.Rules {
+				_, ok := elementRule.Escalations[newElement.EscalationID]
+				if ok {
+					newElement.RuleID = id
+					ruleFound = true
+					break
 				}
+			}
+			if !ruleFound {
+				return fmt.Errorf("rule escalation recipient cannot be mapped to a rule")
+			}
 
-				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(r *rule.EscalationRecipient) bool {
-					return r == nil
-				})
+			escalation := r.Rules[newElement.RuleID].Escalations[newElement.EscalationID]
```
You can use `r.GetRuleEscalation(newElement.EscalationID)` here instead.

oxzi replied:
Unfortunately, `RuntimeConfig.GetRuleEscalation` only returns a `*rule.Escalation`, but not a reference to the `escalation.Rule`. In particular, the Rule ID is necessary to delete this `EscalationRecipient` again, as there is no link to `Escalations` at the `RuntimeConfig`.

https://github.com/Icinga/icinga-notifications/blob/767e4b8ef142c6e0bfd11573ace18576f6811906/internal/config/rule.go#L110-L125

yhabteab replied:
> In particular, the Rule ID is necessary to delete this `EscalationRecipient` again, as there is no link to `Escalations` at the `RuntimeConfig`.

If `RuntimeConfig.GetRuleEscalation` does not return nil, you can simply access the rule ID via `escalation.RuleID`..

yhabteab replied:
IMHO, adding the additional field `RuleID` to `EscalationRecipient` is not even necessary as a recipient has a reference to its escalation and the escalation to its rule.

oxzi replied:
Valid point. Now I am going to directly refer `*Escalation`, easing both the `createFn` and `delteFn`. Thanks :)

yhabteab reviewed (commented):
> You have defined json tags for all ChangedAt and Deleted fields, but seem to ignore them completely in all update function hooks. Is that intentional? If so, then the json tags are probably useless.

I have completed only half of the changes so far! But what about this question?

yhabteab commented on internal/config/schedule.go:
```diff
@@ -1,172 +1,156 @@
-	rotationMembersById := make(map[int64]*recipient.RotationMember)
-	for _, member := range members {
-		memberLogger := r.logger.With(zap.Object("rotation_member", member))
+			rotationId := slices.IndexFunc(schedule.Rotations, func(rotation *recipient.Rotation) bool {
+				return rotation.ID == newElement.RotationID
+			})
+			if rotationId < 0 {
+				return fmt.Errorf("schedule rotation member refers unknown rotation %d", newElement.RotationID)
+			}
```
How can you end up in this situation? Isn't that schedule retrieved based on `newElement.RotationID` in the above loop and handled accordingly if it is nil? I'm also wondering why you added the `ScheduleID` field to `RotationMember` type.

oxzi replied:
This is a case of "this should never happen", but I prefer a helpful error instead of a panic resulting in the next line.

yhabteab replied:
> but I prefer a helpful error instead of a panic resulting in the next line.

As far as I can see, there's no way that could result in a panic here, since you literally searched the `Schedule` via the rotation ID. This IMHO just adds some dead codes that shouldn't be needed at all.

oxzi replied:
Sorry, I failed to get your point earlier as I was just looking at the quoted code section. However, when scrolling out to see the bigger picture, I get it.

oxzi commented on internal/config/contact.go:
```diff
@@ -1,62 +1,54 @@
-	if r.Contacts == nil {
-		r.Contacts = make(map[int64]*recipient.Contact)
-	}
+			contact.Addresses = append(contact.Addresses, newElement)
```
The old code does not really made sense, in my opinion.

https://github.com/Icinga/icinga-notifications/blob/32bf26e3217a64697faab1ce9f8497a00e0a9517/internal/config/contact_address.go#L67-L80

Based on the database schema, the table `contact_address` has an `id` as its primary key and an `UNIQUE` constraint about the tuple `contact_id, type`.

https://github.com/Icinga/icinga-notifications/blob/767e4b8ef142c6e0bfd11573ace18576f6811906/schema/pgsql/schema.sql#L74-L85

The `if` clause you highlighted just checks if this very `contact_address` row - as a `recipient.Address` - is not already part of its referencing `contact` row - as a `recipient.Contact`. However, based on the schema and how the changes are being synchronized, I don't think this "check" is necessary. I would even call it counterproductive.

yhabteab replied:
You are right, but we should not forget to adjust the code when this comment becomes reality. `constraint may be relaxed in the future to support multiple addresses per type `

oxzi commented:
> > You have defined json tags for all ChangedAt and Deleted fields, but seem to ignore them completely in all update function hooks. Is that intentional? If so, then the json tags are probably useless.
> 
> I have completed only half of the changes so far! But what about this question?

@yhabteab: I have already removed the `json` tags as they were, as you mentioned, useless. My initial idea was to refer to them within the JSON dump, but as most other fields are missing such a tag, I removed them here as well.

yhabteab commented:
> > > You have defined json tags for all ChangedAt and Deleted fields, but seem to ignore them completely in all update function hooks. Is that intentional? If so, then the json tags are probably useless.
> > 
> > 
> > I have completed only half of the changes so far! But what about this question?
> 
> @yhabteab: I have already removed the `json` tags as they were, as you mentioned, useless. My initial idea was to refer to them within the JSON dump, but as most other fields are missing such a tag, I removed them here as well.

I mean, okay, you removed the json tag completely and my comment didn't even propose to remove the tags at all, but if you think these fields should also be included in the config dump, you need to update them regularly like the other fields, otherwise just exclude them via `json:"-"`. Currently, they should still appear in the config dump with false-positive values, since both fields are exported.

oxzi commented on internal/timeperiod/timeperiod.go:
```diff
@@ -65,10 +80,29 @@ type Entry struct {
+	return e.Deleted.Valid && e.Deleted.Bool
+}
+
+func (e *Entry) IncrementalValidate() error {
+	return e.Init()
+}
```
Valid point! I've changed it to a Javaesque `IncrementalConfigurableInitAndValidatable`.

oxzi commented on internal/timeperiod/timeperiod.go:
```diff
@@ -77,12 +111,16 @@ func (e *Entry) TableName() string {
 		encoder.AddString("rrule", e.RRule.String)
 	}
+	if e.RotationMemberID.Valid {
+		encoder.AddInt64("rotation-member-id", e.RotationMemberID.Int64)
```
There's both in the code base, but `_` are more prominent.

```
$ grep -Er '(zap|encoder\.Add).*".*_.*"' internal | wc -l
27
$ grep -Er '(zap|encoder\.Add).*".*-.*"' internal | wc -l
17
```

Thus, I am going to change it to `_` and maybe this could be unified later on.

yhabteab replied:
New changes introduced with this PR! :)
```bash
$ grep -Enr '(zap\.|encoder\.).*\(".*-.*"' internal
internal/config/incremental_sync.go:110:                        zap.Int("deleted-elements", countDel),
internal/config/incremental_sync.go:111:                        zap.Int("faulty-elements", countErr),
internal/config/incremental_sync.go:112:                        zap.Int("loaded-elements", countLoad))
internal/config/incremental_sync.go:241:                        zap.Int("faulty-elements", countErr),
internal/config/incremental_sync.go:242:                        zap.Int("deleted-unknown-elements", countDelSkip),
internal/config/incremental_sync.go:243:                        zap.Int("deleted-elements", countDel),
internal/config/incremental_sync.go:244:                        zap.Int("updated-elements", countUpdate),
internal/config/incremental_sync.go:245:                        zap.Int("new-elements", countNew))
```

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,247 @@
+// The RuntimeConfig.configChangeTimestamps map contains the last known timestamp for each BaseT table. Only those
+// elements where the changed_at SQL column is greater than the stored timestamp will be fetched and stored in the
+// temporary RuntimeConfig.configChange ConfigSet. Later on, incrementalApplyPending merges it into the main ConfigSet.
+func incrementalFetch[
+	BaseT any,
+	PK comparable,
+	T interface {
+		*BaseT
+		IncrementalConfigurable[PK]
+	},
+](ctx context.Context, tx *sqlx.Tx, r *RuntimeConfig, changeConfigSetField *map[PK]T) error {
```
I really like this idea, but, unless I am missing some point, this does not work. Your suggestion would add another type parameter `TPtr` which is not a requirement of `T`, but the other way around. Thus, using `*T` as the type would not allow accessing the methods defined by the `IncrementalConfigurable[PK]` interface. The `TPtr` type parameter would be unused and marked as such by the compiler.

To play around with this and to demonstrate, I created a mock-up here: https://go.dev/play/p/3XVTLj3Mjgu

Please feel free to correct me if I have misunderstood your point.

julianbrost replied:
I've missed that problem but it can be worked around by casting to `TPtr` (no type assertion required):
```go
TPtr(t).GetPrimaryKey()
```

Now the question is whether this makes it ugly again.

oxzi replied:
IMHO, this would make it even more complex to comprehend. Then I would keep it as it is.

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,247 @@
+	return nil
+}
+
+// reAddUpdateFnErr is a special non-error which might be expected from incrementalApplyPending's updateFn to signal
+// that the current element should be updated by being deleted through the deleteFn first and added again by the
+// createFn hook function. This implies that both createFn and deleteFn are set and not a nil pointer.
+var reAddUpdateFnErr = errors.New("re-adding by invoking the deletion function followed by the creation function")
```
I renamed it but kept it lowercase as it is not required to export it outside the package's scope.

oxzi commented on internal/config/timeperiod.go:
```diff
@@ -1,100 +1,51 @@
+			}
 
-	r.pending.TimePeriods = nil
+			period.Entries = slices.DeleteFunc(period.Entries, func(entry *timeperiod.Entry) bool {
+				return entry.ID == delElement.ID
+			})
```
I tried not to change code directly outside of this change's scope too much. On the one hand, it would simplify those apply methods where I have to toy around with the `slices` package, but on the other hand, the diff would be bigger and reach more in other packages. Thus, I would prefer to keep it this way.

oxzi commented on internal/config/schedule.go:
```diff
@@ -1,172 +1,156 @@
+			updatedScheduleIds[schedule.ID] = struct{}{}
+			return nil
+		},
+		func(curElement, update *recipient.Rotation) error {
+			curElement.ActualHandoff = update.ActualHandoff
+			curElement.Priority = update.Priority
+			curElement.Name = update.Name
+			return nil
+		},
```
Based on the whole architecture, especially in the web part, I would not expect a rotation to switch schedules, even if technically possible. But yes, valid point. I have added a check for this case, resulting in re-adding the rotation, same as for `recipient.Address`. I did the same for `rule.Escalation`.

oxzi commented on internal/config/schedule.go:
```diff
@@ -1,172 +1,156 @@
+		func(newElement *recipient.RotationMember) error {
+			newElement.TimePeriodEntries = make(map[int64]*timeperiod.Entry)
+
+			var schedule *recipient.Schedule
+			for id, tmpSchedule := range r.Schedules {
+				ok := slices.ContainsFunc(tmpSchedule.Rotations, func(rotation *recipient.Rotation) bool {
+					return rotation.ID == newElement.RotationID
+				})
+				if ok {
+					newElement.ScheduleID = id
+					schedule = tmpSchedule
+					break
+				}
+			}
+			if schedule == nil {
+				return fmt.Errorf("schedule rotation member cannot be mapped to a schedule")
+			}
```
Yes.. this got a bit messed up in between, especially after I just refactored it again in spirit of <https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1657337611>.

However, now I am accessing the unexported intermediate table struct fields. Should be cleaner now.

oxzi commented:
I have tried to work on all of your comments and closed those discussions I though I have properly addressed. Please feel free to reopen those, if you think otherwise.

Furthermore, I made sure that the `changed_at` value is also always being updated, as @yhabteab pointed out.

Please take another look and sorry for the broken diff as I needed to rebase the code :/

julianbrost commented on schema/pgsql/upgrades/032.sql:
```diff
@@ -0,0 +1,94 @@
+
+CREATE INDEX idx_rule_changed_at ON rule(changed_at);
+
+UPDATE rule SET deleted = 'y' WHERE is_active = 'n';
+ALTER TABLE rule DROP COLUMN is_active;
```
That seems to be new now?

oxzi replied:
As @sukhwinder33445 is implementing the web counterpart, he wrote me that `is_active` is getting useless after rules can be removed with `deleted`. It seems like web does not need it anymore now.

julianbrost commented on schema/pgsql/schema.sql:
```diff
@@ -48,55 +48,84 @@ CREATE TABLE channel (
     address text NOT NULL, -- phone number, email address, ...
 
-    CONSTRAINT pk_contact_address PRIMARY KEY (id),
-    UNIQUE (contact_id, type) -- constraint may be relaxed in the future to support multiple addresses per type
```
And dropping this constraint also looks new. Why is it removed?

oxzi replied:
This is also based on a discussion with @sukhwinder33445.

The uniqueness constraint forbids re-adding another contact address of a previously deleted - via `deleted = 'y'` - type. When setting the `type` column to `NULL`, as did elsewhere, this would mess up history mentioning deleted entries. As there was still no consensus if we want to support showing deleted entries, the goal was not to purge history.

There was also the idea of having a conditional `UNIQUE` constraint, similar to your discussion in <https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1660678059>.

After going back and forth, we decided to just drop the `UNIQUE` constraint. However, I am very open for other suggestions.

julianbrost replied:
> When setting the `type` column to `NULL`, as did elsewhere, this would mess up history mentioning deleted entries.

Does the history currently show the addresses? Should it even do that (that might even bringt up privacy concerns, "John Doe was notified via SMS" and "John Doe was notified by SMS to 0118999881[...]" are quite different things)?

oxzi replied:
Data protection and privacy is even another issue. For example, to comply with the GDPR, it must generally be possible to remove personal data, even if Art. 17 GDPR allows some exceptions. However, law is definitely not my domain.

Back on the technical ground, I would argue in favor of lifting this constraint. It was already marked with a comment to be potentially removed and I don't see any current limitation in its favor. Next to enabling the `deleted = 'y'` logic required here, it would allow, i.e., multiple phone numbers per contact. To stay with your example, some contact might get notified both on their old and new phone number during transition periods, e.g., getting messages both to the old 999 and the new 0118 999 881 999 119 7253.

yhabteab commented on internal/config/rule.go:
```diff
@@ -1,195 +1,124 @@
+		},
+		nil,
+		func(delElement *rule.EscalationRecipient) error {
+			delElement.Escalation.Recipients = slices.DeleteFunc(
+				delElement.Escalation.Recipients,
+				func(recipient *rule.EscalationRecipient) bool {
+					return recipient.EscalationID == delElement.EscalationID
+				})
```
```suggestion
			escalation := r.GetRuleEscalation(delElement.EscalationID)
			if escalation != nil {
				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(recipient *rule.EscalationRecipient) bool {
					return recipient.EscalationID == delElement.EscalationID
				})
			}
```

There is absolutely no reason to introduce an escalation field to the recipients type.

oxzi replied:
Totally valid point. The extra struct field seemed necessary in a way earlier state of this PR.

oxzi commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,251 @@
+// removeAndAddInstead is a special non-error which might be expected from incrementalApplyPending's updateFn to signal
+// that the current element should be updated by being deleted through the deleteFn first and added again by the
+// createFn hook function. This implies that both createFn and deleteFn are set and not a nil pointer.
+var removeAndAddInstead = errors.New("re-adding by invoking the deletion function followed by the creation function")
```
This name was based on @julianbrost's <https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1662578483>.

As I stated in my answer to this comment before, I don't see any reason to export this variable outside of the package scope as all synchronization functions are also only package-local.

I have renamed the variable to `errRemoveAndAddInstead`.

oxzi commented on internal/config/rule.go:
```diff
@@ -1,195 +1,124 @@
+	// ID field. If targetRule equals referenceRule, it can be used for initialization, otherwise it updates.
+	ruleSetTimePeriod := func(targetRule, referenceRule *rule.Rule) error {
+		if referenceRule.TimePeriodID.Valid {
+			if tp, ok := r.TimePeriods[referenceRule.TimePeriodID.Int64]; !ok {
+				return fmt.Errorf("rule refers unknown time period %d", referenceRule.TimePeriodID.Int64)
+			} else {
+				targetRule.TimePeriod = tp
```
The code within this anonymous function requires access to the current `*RuntimeConfig`. The `IncrementalInitAndValidate` method from the `IncrementalConfigurableInitAndValidatable` interface is called by `incrementalFetch`. When being called, no new state has been synchronized to the current `*RuntimeConfig`, resulting the check to fail.

The `IncrementalConfigurableInitAndValidatable` interface is more like a preflight check for each element by itself. The check in question, however, compares against further state.

yhabteab replied:
Sorry! I wasn't aware of this that match.

julianbrost commented on internal/config/rule.go:
```diff
@@ -1,195 +1,127 @@
+		r,
+		&r.ruleEscalationRecipients, &r.configChange.ruleEscalationRecipients,
+		func(newElement *rule.EscalationRecipient) error {
+			ok := false
+			if newElement.ContactID.Valid {
+				newElement.Recipient, ok = r.Contacts[newElement.ContactID.Int64]
+			} else if newElement.GroupID.Valid {
+				newElement.Recipient, ok = r.Groups[newElement.GroupID.Int64]
+			} else if newElement.ScheduleID.Valid {
+				newElement.Recipient, ok = r.Schedules[newElement.ScheduleID.Int64]
+			}
```
`EscalationRecipient` embeds `recipient.Key`, so you should be able to use the following function here right away:
https://github.com/Icinga/icinga-notifications/blob/32bf26e3217a64697faab1ce9f8497a00e0a9517/internal/config/runtime.go#L112-L132

julianbrost reviewed (commented):
I'm not done with looking at everything yet, but there's already enough so that you might have a look at it.

oxzi commented on internal/incident/incident.go:
```diff
@@ -475,10 +471,6 @@ func (i *Incident) evaluateEscalations(eventTime time.Time) ([]*rule.Escalation,
 	for rID := range i.Rules {
 		r := i.runtimeConfig.Rules[rID]
 
-		if r == nil || !r.IsActive.Valid || !r.IsActive.Bool {
-			continue
-		}
```
I hoped so, but to be sure I have added a safety check.

oxzi commented on internal/config/schedule.go:
```diff
@@ -1,172 +1,133 @@
 
-		if member == nil {
-			r.logger.Warnw("ignoring entry for unknown rotation_member_id",
-				zap.Int64("timeperiod_entry_id", entry.ID),
-				zap.Int64("timeperiod_id", entry.TimePeriodID))
+		rotationMember, ok := r.scheduleRotationMembers[timeperiodEntry.RotationMemberID.Int64]
+		if !ok {
+			r.logger.Errorw("Time period entry refers unknown rotation member", zap.Inline(timeperiodEntry))
 			continue
 		}
 
-		err := entry.Init()
+		err := timeperiodEntry.Init()
 		if err != nil {
-			r.logger.Warnw("ignoring time period entry",
-				zap.Object("entry", entry),
-				zap.Error(err))
+			// This shouldn't happen as every element should already be initialized.
+			r.logger.Errorw("Cannot initialize time period entry", zap.Inline(timeperiodEntry), zap.Error(err))
 			continue
 		}
 
-		member.TimePeriodEntries[entry.ID] = entry
+		rotationMember.TimePeriodEntries[timeperiodEntry.ID] = timeperiodEntry
 	}
```
I have reordered the synchronization of the schedule- and timeperiod-related entries, allowing linking and unlinking in `r.timePeriodEntries`' `createFn` and `deleteFn`.

oxzi commented on internal/config/rule.go:
```diff
@@ -1,195 +1,127 @@
-	}
-
-	for _, recipient := range recipients {
-		recipientLogger := r.logger.With(
-			zap.Int64("id", recipient.ID),
-			zap.Int64("escalation_id", recipient.EscalationID),
-			zap.Int64("channel_id", recipient.ChannelID.Int64))
-
-		escalation := escalationsByID[recipient.EscalationID]
-		if escalation == nil {
-			recipientLogger.Warnw("ignoring recipient for unknown escalation")
 		} else {
-			escalation.Recipients = append(escalation.Recipients, recipient)
-			recipientLogger.Debugw("loaded escalation recipient config")
+			targetRule.TimePeriod = nil
 		}
+		targetRule.TimePeriodID = referenceRule.TimePeriodID
+		return nil
 	}
```
I have inlined the helper function again. Doing so eased the function a lot, as the special cases are only necessary at one calling point.

oxzi commented on schema/pgsql/schema.sql:
```diff
@@ -258,25 +315,35 @@ CREATE TABLE rule (
 
-    UNIQUE (rule_id, position),
+    UNIQUE (rule_id, position), -- column must be NULLed for deletion via "deleted = 'y'"
+    CHECK (deleted = 'y' OR position IS NOT NULL),
     CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL))
```
I have both reordered the CHECK constraints and named them. However, in the schema upgrade this required two distinct `ALTER TABLE ` queries as `RENAME CONSTRAINT` is not a "normal" `action` to be chained, <https://www.postgresql.org/docs/current/sql-altertable.html>.

julianbrost commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,256 @@
+	appliedChanges bool,
+) {
+	startTime := time.Now()
+	tableName := database.TableName(*new(T))
```
Hm, that looks pretty confusing. If I didn't get something wrong, `*new(T)` should do exactly the same as `T(nil)` here.

Even if we don't change which of the type parameters is used throughout the function, maybe we should rename them to `T` (from `BaseT`) and `Tptr` (from `T`) so that it's at least somewhat clear that one of them is actually a pointer type. At least `*new(Tptr)` would look more suspect to me that it currently does.

oxzi replied:
Technically, `new(BaseT)` should also work. However, I found this confusing as otherwise only `T` is being used. I also disliked the idea of introducing a new variable to be used just here. Furthermore, calling `new` is useless anyway. Somehow, my brain did not end up at `T(nil)`, which I now used.

I started to take a liking for the current generic type names, being constructed from somehow descriptive names to the good old `T`, indicating that this is the generic type to use.

julianbrost commented on internal/utils/flag.go:
I'm not a huge fan of this type. I think it should be easier to just use simple bool operations instead like this:

```go
fooChanged := incrementalApplyPending(...)

barChanged := incrementalApplyPending(...)

return fooChanged || barChanged
```

```go
changed := false
for _, f := range applyFns {
	if f() {
		changed = true
	}
}
return changed
```

It's just more obvious what's going on compared to `flag.Update(...)` where you have to refer to the documentation to follow. And these operations are very simple, you aren't even saving code.

julianbrost replied:
Or, another possibility: instead of returning, `incrementalApplyPending()` could just set a new attribute `RuntimeConfig.unverifiedChanges` to true. Then you wouldn't have to pass that around all the time and the purpose of that function is to modify `*RuntimeConfig`, so just having a flag there doesn't sound wrong to me at first glance.

oxzi replied:
I went with your second option, introducing another variable in the `RuntimeConfig`. At first, I thought about this approach, but it felt untidy in my head. However, after what I ended up here, this is a lot cleaner.

julianbrost commented on internal/config/timeperiod.go:
```diff
@@ -1,100 +1,67 @@
+		r,
+		&r.TimePeriods, &r.configChange.TimePeriods,
+		nil,
+		func(element, update *timeperiod.TimePeriod) error {
+			element.Name = update.Name
+			return nil
+		},
```
This misses the `ChangedAt` update.

julianbrost commented on internal/config/schedule.go:
```diff
@@ -1,172 +1,118 @@
+		r,
+		&r.Schedules, &r.configChange.Schedules,
+		nil,
+		func(curElement, update *recipient.Schedule) error {
+			curElement.Name = update.Name
+			return nil
+		},
```
This one also misses the `ChangedAt` update.

julianbrost commented on schema/pgsql/schema.sql:
```diff
@@ -141,10 +185,13 @@ CREATE TABLE rotation_member (
     UNIQUE (rotation_id, contact_id),
     UNIQUE (rotation_id, contactgroup_id),
     CHECK (num_nonnulls(contact_id, contactgroup_id) = 1),
+    CHECK (deleted = 'y' OR position IS NOT NULL),
```
This still creates constraints named `rotation_member_check` and `rotation_member_check1`, so better name them here as well.

julianbrost commented on schema/pgsql/schema.sql:
```diff
@@ -258,25 +315,35 @@ CREATE TABLE rule (
-    UNIQUE (rule_id, position),
-    CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL))
+    UNIQUE (rule_id, position), -- column must be NULLed for deletion via "deleted = 'y'"
+    CONSTRAINT ck_rule_escalation_condition_or_fallback_for CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL)),
+    CONSTRAINT ck_rule_escalation_deleted_or_position_null CHECK (deleted = 'y' OR position IS NOT NULL)
```
I'd use more descriptive names for the constraints. Maybe something like "fallback has no condition" or "active row has position".

yhabteab reviewed (commented):
> There's both in the code base, but `_` are more prominent.
> 
> ```
> $ grep -Er '(zap|encoder\.Add).*".*_.*"' internal | wc -l
> 27
> $ grep -Er '(zap|encoder\.Add).*".*-.*"' internal | wc -l
> 17
> ```
> 
> Thus, I am going to change it to `_` and maybe this could be unified later on.

New changes introduced with this PR! :)
```bash
$ grep -Enr '(zap\.|encoder\.).*\(".*-.*"' internal
internal/config/incremental_sync.go:110:                        zap.Int("deleted-elements", countDel),
internal/config/incremental_sync.go:111:                        zap.Int("faulty-elements", countErr),
internal/config/incremental_sync.go:112:                        zap.Int("loaded-elements", countLoad))
internal/config/incremental_sync.go:244:                        zap.Int("faulty-elements", countErr),
internal/config/incremental_sync.go:245:                        zap.Int("deleted-unknown-elements", countDelSkip),
internal/config/incremental_sync.go:246:                        zap.Int("deleted-elements", countDel),
internal/config/incremental_sync.go:247:                        zap.Int("updated-elements", countUpdate),
internal/config/incremental_sync.go:248:                        zap.Int("new-elements", countNew))
```

yhabteab commented on internal/rule/escalation.go:
```diff
@@ -93,13 +115,24 @@ func (e *Escalation) TableName() string {
+	if r.ChannelID.Valid {
+		encoder.AddInt64("channel_id", r.ChannelID.Int64)
+	}
+	return nil
```
Now you are completely overriding the parent method. Please add `return r.Key.MarshalLogObject(encoder)` perhaps at the end instead of returning nil.

julianbrost reviewed (changes requested):
The logging needs some fine-tuning, especially at startup, too much like the following is logged:

```
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 19, "new": {"contact_id": 19, "name": "Bernd"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 21, "new": {"contact_id": 21, "name": "Blerim"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 20, "new": {"contact_id": 20, "name": "Eric"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 2, "new": {"contact_id": 2, "name": "Julian Brost"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 1, "new": {"contact_id": 1, "name": "Icinga Admin"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 23, "new": {"contact_id": 23, "name": "Rotation Recipient Weekday 10-14"}}
2024-07-09T14:53:29.868Z    INFO    runtime-updates Created new configuration element   {"table": "contact", "id": 24, "new": {"contact_id": 24, "name": "Rotation Weekday 8-20"}}
```

Additionally, there are some confusing warnings logged when the first refresh after the initial load happens:

```
2024-07-09T14:53:30.874Z    WARN    runtime-updates Skipping unknown marked as deleted configuration element    {"table": "contact", "id": 11}
2024-07-09T14:53:30.874Z    WARN    runtime-updates Skipping unknown marked as deleted configuration element    {"table": "contact", "id": 22}
2024-07-09T14:53:30.874Z    WARN    runtime-updates Skipping unknown marked as deleted configuration element    {"table": "contact", "id": 9}
2024-07-09T14:53:30.874Z    WARN    runtime-updates Skipping unknown marked as deleted configuration element    {"table": "contact", "id": 10}
2024-07-09T14:53:30.874Z    WARN    runtime-updates Skipping unknown marked as deleted configuration element    {"table": "contact", "id": 25}
```

These appear because the timestamps are only set to `MAX(changed_at) WHERE deleted = 'n'` after the initial load because deleted rows were skipped, so if rows were deleted after the other change, they will be processed on the first update. Doing that doesn't really break anything, the logging at level WARNING is just very strange.

oxzi commented:
I have changed the detection of initially deleted elements to cover this special case of the latest change being a deletion. Regarding the log entry flood of created elements, I reconsidered the current log levels and have degraded everything except errors and the summary after changes to DEBUG.

yhabteab commented on internal/incident/incident.go:
```diff
@@ -474,8 +470,10 @@ func (i *Incident) evaluateEscalations(eventTime time.Time) ([]*rule.Escalation,
-		if r == nil || !r.IsActive.Valid || !r.IsActive.Bool {
+		if r == nil {
+			i.logger.Warnw("Incident refers unknown rule",
+				zap.String("incident", i.String()),
```
You don't have to add `incident` here, as `i.logger` should already contain that key.

oxzi replied:
I'm honestly unsure. It seems like `i.logger` only gets populated with `zap.String("incident", i.String())` during `ProcessEvent` if this incident is seen as "new", defined by having a zero `started_at` value.

https://github.com/Icinga/icinga-notifications/blob/853992e30328b3292dbcf018df95fe9bbcbe3a31/internal/incident/incident.go#L149-L157

I mean yes, I guess it can be removed at this position. But does it might occur that some Incidents are without this field, e.g., for reevaluation?

julianbrost replied:
That sounds like if there are cases where the incident logger doesn't contain metadata fields for the incident, that's a bug.

yhabteab replied:
> But does it might occur that some Incidents are without this field, e.g., for reevaluation?

Nope!
https://github.com/Icinga/icinga-notifications/blob/e954e4059911d3bd9b6d6720a4de13f65609f2e8/internal/incident/incidents.go#L124-L125

oxzi replied:
Thanks @yhabteab. I somehow missed that one.

yhabteab commented on schema/pgsql/schema.sql:
```diff
@@ -48,55 +48,84 @@ CREATE TABLE channel (
 );
 
+CREATE INDEX idx_channel_changed_at ON channel(changed_at);
+
 CREATE TABLE contact (
     id bigserial,
     full_name citext NOT NULL,
-    username citext, -- reference to web user
+    username citext, -- reference web user
```
This change doesn't make sense to me. Is this a accidental change?

oxzi replied:
Actually no. This was the result of multiple changes :upside_down_face:

yhabteab reviewed (dismissed):
I have tested (hopefully 🤞 all use cases) and works perfectly for me!

I can't reproduce the error from yesterday, but even then I don't think it's a bug in the GO code, but rather in the web part.

I was only able to get the error again by manually editing some deleted rotation members in the database.
```bash
2024-07-10T11:00:42.886+0200    ERROR   runtime-updates Creating configuration element failed   {"table": "rotation_member", "id": 30, "new": {"id": 30, "rotation_id": 16, "contact_group_id": 1}, "error": "creation callback error, schedule rotation member refers unknown rotation 16"}
```

julianbrost commented on schema/pgsql/schema.sql:
```diff
@@ -258,25 +316,35 @@ CREATE TABLE rule (
-    UNIQUE (rule_id, position),
-    CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL))
+    UNIQUE (rule_id, position), -- column must be NULLed for deletion via "deleted = 'y'"
+    CONSTRAINT ck_rule_escalation_fallback_has_no_fallback CHECK (NOT (condition IS NOT NULL AND fallback_for IS NOT NULL)),
```
Fallback has no fallback? One of these words should be condition :sweat_smile:

oxzi replied:
No Thoughts Head Empty.

julianbrost commented on schema/pgsql/schema.sql:
```diff
@@ -105,32 +134,47 @@ CREATE TABLE rotation (
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    deleted boolenum NOT NULL DEFAULT 'n',
+
+    -- Each schedule can only have one rotation with a given priority starting at a given date.
+    -- Column must be NULLed for deletion via "deleted = 'y'".
     UNIQUE (schedule_id, priority, first_handoff),
```
> Column must be NULLed

It's unclear which column this refers to. Also, this information is given inconsistently throughout the file, sometimes it's a comment on the constraint, sometimes on the column. Maybe this can be unified as well.

oxzi replied:
I have unified the comments, put them above the `UNIQUE` line describing the reason and which columns must be NULLed.

julianbrost commented on schema/pgsql/upgrades/032.sql:
```diff
@@ -0,0 +1,100 @@
+    ALTER COLUMN position DROP NOT NULL,
+    ADD COLUMN changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    ADD COLUMN deleted boolenum NOT NULL DEFAULT 'n',
+    ADD CONSTRAINT ck_rule_escalation_fallback_has_no_fallback CHECK (deleted = 'y' OR position IS NOT NULL);
```
The names of the constraints are mixed up here.

julianbrost commented on internal/config/incremental_sync.go:
```diff
@@ -0,0 +1,259 @@
+	stmtLogger := r.logger.With(zap.String("table", tableName))
+
+	stmt := r.db.BuildSelectStmt(typePtr, typePtr)
+	stmtArgs := []any{}
```
```suggestion
	var stmtArgs []any
```

No need to initialize the slice here, `nil` works just fine and doesn't allocate[^1] something that's thrown away in most cases anyways and is not necessary in the remaining case.

[^1]: Maybe the compiler is smart enough to optimize it away, but I don't see a downside of just using `nil` here and that wouldn't even need a smart compiler.

julianbrost commented:
> I was only able to get the error again by manually editing some deleted rotation members in the database.
> 
> ```shell
> 2024-07-10T11:00:42.886+0200    ERROR   runtime-updates Creating configuration element failed   {"table": "rotation_member", "id": 30, "new": {"id": 30, "rotation_id": 16, "contact_group_id": 1}, "error": "creation callback error, schedule rotation member refers unknown rotation 16"}
> ```

Did editing involve setting `deleted = 'n'` again on that row? If yes, that's not something that's really intended to be done (or if you do, you'd at least have to undelete all other dependencies as well). Otherwise, that might be a bug as you shouldn't be able to end up in the "Creating configuration element" case with a `deleted = 'y'` row.

oxzi commented:
> > I was only able to get the error again by manually editing some deleted rotation members in the database.
> > ```shell
> > 2024-07-10T11:00:42.886+0200    ERROR   runtime-updates Creating configuration element failed   {"table": "rotation_member", "id": 30, "new": {"id": 30, "rotation_id": 16, "contact_group_id": 1}, "error": "creation callback error, schedule rotation member refers unknown rotation 16"}
> > ```
> 
> Did editing involve setting `deleted = 'n'` again on that row? If yes, that's not something that's really intended to be done (or if you do, you'd at least have to undelete all other dependencies as well). Otherwise, that might be a bug as you shouldn't be able to end up in the "Creating configuration element" case with a `deleted = 'y'` row.

The actual error from the "creation callback [states, that the] schedule rotation member refers unknown rotation 16". The creation callback cannot[^1] be called upon an `deleted = 'y'` element. Thus, I would guess that you have manually revived a Rotation Member belonging to a still deleted Rotation? Otherwise, something would be wrong.

[^1]: Should not. I really hope so, at least.

julianbrost commented on schema/pgsql/upgrades/032.sql:
```diff
@@ -0,0 +1,100 @@
+    ALTER COLUMN position DROP NOT NULL,
+    ADD COLUMN changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    ADD COLUMN deleted boolenum NOT NULL DEFAULT 'n',
+    ADD CONSTRAINT ck_rotation_member_non_deleted_needs_position CHECK (deleted = 'y' OR position IS NOT NULL),
+
+CREATE INDEX idx_rule_escalation_changed_at ON rule_escalation(changed_at);
```
```
ERROR:  syntax error at or near "CREATE"
LINE 6: CREATE INDEX idx_rule_escalation_changed_at ON rule_escalati...
        ^
```

```suggestion
    ADD CONSTRAINT ck_rotation_member_non_deleted_needs_position CHECK (deleted = 'y' OR position IS NOT NULL);

CREATE INDEX idx_rule_escalation_changed_at ON rule_escalation(changed_at);
```

oxzi replied:
😳

julianbrost commented on schema/pgsql/upgrades/032.sql:
```diff
@@ -0,0 +1,100 @@
+ALTER TABLE rule_escalation
+    RENAME CONSTRAINT rule_escalation_check TO ck_rule_escalation_not_both_condition_and_fallback_for;
+
+ALTER TABLE rule_escalation
+    ALTER COLUMN position DROP NOT NULL,
+    ADD COLUMN changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    ADD COLUMN deleted boolenum NOT NULL DEFAULT 'n',
+    ADD CONSTRAINT ck_rotation_member_non_deleted_needs_position CHECK (deleted = 'y' OR position IS NOT NULL),
```
Still not quite the right constraint name :see_no_evil:

Diff between freshly imported and upgraded schema:
```diff
 CREATE TABLE public.rule_escalation (
     id bigint NOT NULL,
     rule_id bigint NOT NULL,
     "position" integer,
     condition text,
     name public.citext,
     fallback_for bigint,
     changed_at bigint DEFAULT (EXTRACT(epoch FROM now()) * (1000)::numeric) NOT NULL,
     deleted public.boolenum DEFAULT 'n'::public.boolenum NOT NULL,
-    CONSTRAINT ck_rule_escalation_non_deleted_needs_position CHECK (((deleted = 'y'::public.boolenum) OR ("position" IS NOT NULL))),
+    CONSTRAINT ck_rotation_member_non_deleted_needs_position CHECK (((deleted = 'y'::public.boolenum) OR ("position" IS NOT NULL))),
     CONSTRAINT ck_rule_escalation_not_both_condition_and_fallback_for CHECK ((NOT ((condition IS NOT NULL) AND (fallback_for IS NOT NULL))))
 );
```

oxzi replied:
🫠

yhabteab commented:
> Thus, I would guess that you have manually revived a Rotation Member belonging to a still deleted Rotation?

That's exactly what I mean when I wrote "by manually editing". Btw. just updated my local branch for the web part and now it is allowing me to delete event rules and noticed this noise.

```bash
2024-07-10T15:34:15.325+0200    WARN    incident        Incident refers unknown rule    {"object": "dummy-7!onesecond-0", "incident": "#125", "rule_id": 2}
```

If you don't drop that faulty rule from `incident#Rules`, it will always complain about that, and I'm not sure how to handle that properly. Sorry, I didn't notice that previously!

oxzi commented:
> Btw. just updated my local branch for the web part and now it is allowing me to delete event rules and noticed this noise.
> 
> ```shell
> 2024-07-10T15:34:15.325+0200    WARN    incident        Incident refers unknown rule    {"object": "dummy-7!onesecond-0", "incident": "#125", "rule_id": 2}
> ```
> 
> If you don't drop that faulty rule from `incident#Rules`, it will always complain about that, and I'm not sure how to handle that properly. Sorry, I didn't notice that previously!

Good catch that I was able to reproduce.

There are a few relationship relations between the `incident` relation - not directly addressed by this PR as it is no configuration - and relations being touched by this PR.
- `rule` via `incident_rule`
- the three relations `contact`, `contactgroup` and `schedule` via `incident_contact`
- `rule_escalation` via `incident_rule_escalation_state`

I am currently taking a look how to reflect deletions from the configuration tables there.

Al2Klimov commented on schema/pgsql/schema.sql:
```diff
@@ -258,25 +320,38 @@ CREATE TABLE rule (
     object_filter text,
-    is_active boolenum NOT NULL DEFAULT 'y',
+
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
```
I'm afraid this isn't portable to MySQL (like https://github.com/Icinga/icinga-notifications/pull/203#discussion_r1618987098).

Al2Klimov replied:
Precisely speaking, we have to provide \*nix ts by ourselves, just like https://github.com/Icinga/icinga-notifications/pull/215/files#diff-e9b5c34a7eeb261757ec19f5cd8fdd7ade8b32d33cbaf742262f967a261da59eR353

julianbrost replied:
What is the problem here? From a quick test, this seems to work fine except that the snippet I copy-pasted from Icinga DB to get a millisecond timestamp (`UNIX_TIMESTAMP() * 1000`) only seems to produce second precision, but that should hopefully be fixable with a different expression.

```
MariaDB [jbrost]> create table default_test (t bigint default (UNIX_TIMESTAMP() * 1000));
Query OK, 0 rows affected (0.015 sec)

MariaDB [jbrost]> insert into default_test () values ();
Query OK, 1 row affected (0.010 sec)

MariaDB [jbrost]> insert into default_test () values ();
Query OK, 1 row affected (0.003 sec)

MariaDB [jbrost]> insert into default_test () values ();
Query OK, 1 row affected (0.009 sec)

MariaDB [jbrost]> select * from default_test;
+---------------+
| t             |
+---------------+
| 1720689757000 |
| 1720689760000 |
| 1720689765000 |
+---------------+
3 rows in set (0.001 sec)
```

Al2Klimov replied:
Sorry, forgot to mention that OLD versions, like Maria 10.1, are the problem:

```
Server version: 10.1.48-MariaDB-1~bionic mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [idb]> create table default_test (t bigint default (UNIX_TIMESTAMP() * 1000));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '(UNIX_TIMESTAMP() * 1000))' at line 1
```

Al2Klimov replied:
FWIW on AL 2 the MariaDB version is even lower.

yhabteab replied:
> FWIW on AL 2 the MariaDB version is even lower.

I wouldn't use that as an argument, as this OS ships mariadb-5.5, which is something we aren't ever going to support. The default mariadb version on SLES 12.5 is `10.2`, so if such defaults work with that version, I would just drop support for `10.1`.

Al2Klimov replied:
```
Server version: 10.2.44-MariaDB-1:10.2.44+maria~bionic mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [idb]> create table default_test (t bigint default (UNIX_TIMESTAMP() * 1000));
Query OK, 0 rows affected (0,021 sec)
```

(RHEL 8 has v10.3 btw.)

julianbrost replied:
@Al2Klimov Which of the versions currently mentioned in the documentation (MySQL (≥5.5), MariaDB (≥10.1)) would no longer be supported if we added that default? So is Maria DB 10.1 the only thing we'd lose?

https://github.com/Icinga/icinga-notifications/blob/f369a116868e8189f321c8ef579b190a3647636a/doc/02-Installation.md#L24

Al2Klimov replied:
We'd also lose MySQL 5.7, I guess:

```
Server version: 5.7.44 MySQL Community Server (GPL)

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MySQL [idb]> create table default_test (t bigint default (UNIX_TIMESTAMP() * 1000));
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(UNIX_TIMESTAMP() * 1000))' at line 1
```

```
Server version: 8.0.38 MySQL Community Server - GPL

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MySQL [idb]> create table default_test (t bigint default (UNIX_TIMESTAMP() * 1000));
Query OK, 0 rows affected (0,067 sec)
```

oxzi replied:
I just checked the available MariaDB version for all of our build targets:

- Debian
    - 11: 10.5.23
    - 12: 10.11.6
- Ubuntu
    - 20.04: 10.3.39
    - 22.04: 10.6.18
    - 23.04: 10.11.2
    - 23.10: 10.11.8
    - 24.04: 10.11.8
- Amazon Linux
    - 2: 5.5.68
    - 2023: 10.5
- Fedora
    - 37: 10.5.23
    - 38: 10.5.23
    - 39: 10.5.25
    - 40: 10.11.8
- openSUSE
    - 15.4: 10.6.15
    - 15.5: 10.6.18
    - 15.6: 10.11.8
- RHEL (versions from Rocky because reasons)
    - 8: 10.11.6
    - 9: 10.11.6
- SLES:
    - 12.5: 10.2.44
    - 15.3: 10.5.8
    - 15.4: 10.6.18
    - 15.5: 10.6.18
    - 15.6: 10.11.8

@Al2Klimov: You mentioned MariaDB 10.1 as a lower bound. Thus, only Amazon Linux 2 with 5.5.68 should be an issue.

oxzi replied:
We changed the schema and have dropped the default. @sukhwinder33445 has already implemented it on the :spider_web: side.

oxzi replied:
https://github.com/Icinga/icinga-notifications/pull/191#issuecomment-2222885560

oxzi commented:
Following new fancy :sparkles: changes were introduced:
- Incidents ignore missing references :see_no_evil: 
- The default `changed_at` was dropped as requested https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1673643514 to support older MariaDB and MySQL versions :older_man: 
- All CI checks are green :heavy_check_mark:
